_posts/2020-11-30-dataops-dynamodb-to-snowflake.md

```diff
@@ -0,0 +1,378 @@
+
+* additions to DynamoDB row attributes (i.e. schema) will not break these views as they only read what they are looking for
+* incorporating new fields is as simple as re-creating the view with the new field specified
+* Casting errors or new data attributes can be discovered from the original raw data views
+
+### Visualising The Data
+
+Having gone to the trouble to make our data available in Snowflake, we may as well show it off with a small chart.
+
+Let's make a table that joins the datasets on the `event_date` field:
+
+```sql
+CREATE OR REPLACE TABLE materialized_longterm_weather_stats AS
+  SELECT 
+      s.event_date,
+      x.daily_global_solar_exposure_MJ_per_m2,
+      r.rainfall_mm,
+      r.rainfall_period_days,
+      t.temp_max_c,
+      s.sunshine_hrs,
+      s.time_0900_wind_direction,
+      s.time_0900_wind_speed_kmh
+  FROM weather_summary s
+  JOIN weather_rainfall r ON r.event_date=s.event_date
+  JOIN weather_solar_exposure x ON x.event_date=s.event_date
+  JOIN weather_tempmax t ON t.event_date=s.event_date
+```
+
+Then simply run a query to visualise the data using Snowflake's built in data exploration tool, [SnowSight](https://docs.snowflake.com/en/user-guide/ui-snowsight.html): 
+
+![Screenshot of a graph created by Snowflake's Snowsight Data Exploration UI](/img/blog/dynamodb-to-snowflake/snowsight.png)
+
+## Performance and DataOps
+
+### Performance Considerations: When To Ingest
+
+You will notice in these examples we have not actually ingested any data to Snowflake - we have simply wrapped schema around data that resides on our own S3 bucket.
+
+Snowflake does a great job of caching results from staged data to improve performance, but as soon as your Virtual Warehouse spins down, it will have to re-read that data from S3. This can be both slow and costly, when you have a significant volume of data.
+
+There are a couple of options available
+- Ingest the data at some point by running a `COPY INTO` or *CTAS* statement
+- Create a [Materialized View](https://docs.snowflake.com/en/user-guide/views-materialized.html)
+
+Both of these will be effective though there are complications and cost considerations with materialized views that often make it simpler to `COPY INTO` a new table.
+
+### DataOps
+
+When working with production systems, you don't want to be creating views in your SQL workbench.
+
+* Use a [DataOps Pipeline](https://mechanicalrock.github.io/2020/11/03/snowflake-dataOps.html) to setup and manage your Cloud Data Warehouse.
+* Materialize and Transform your data with a tool like [DBT](https://www.getdbt.com/) once it is available in your warehouse.
+
+More on that in a future blog post!
+
+## Conclusion
+
+In this article we have demonstrated 
+
+* How simple exporting DynamoDB tables to S3 can be with the new export feature
+* How to work with staged files using Snowflake
+* How to normalize semi-structured data into relational views in Snowflake
+
```

Comment:
Suggestion; I think the article is long enough to consider breaking up into a series of posts?

Reply:
how would you structure the series of posts in terms of content?

Reply:
I think the dotted list you put there could work. An article to look at the DDB exporting to S3, then another to do ingestion and semi-structured querying

Don't feel obliged, it's great as-is too, so probably comes down to your preference.

Reply:
I think the post is definitely on the technical side, but I see no point in splitting it up. If we can hammer home the reason ~ why ~ you would want to do the whole process, the level of detail shown for the complete process is valuable